state_dict saving

julianmack opened the pull request:
Small PR updating the ``seq_to_seq`` state dict to include the learning-rate scheduler and optimizer state so that crashed training runs can be resumed w/o disruption. 

I've also added a helper function that, given a state_dict filepath, loads the dict into a ``seq_to_seq`` instance and returns the training state necessary for the `CallbackHandler`. It needs this state so that the `Saver` callback can save with naming convention `state_dict_x` and the `TensorBoardLogger` can resume from the same step. 
 
I have not added tests for this functionality here - but I have added one in the lr scheduler PR: https://github.com/MyrtleSoftware/myrtlespeech/pull/22 (which is blocked by the `rnnt` PR). The reason for this is that the state_dict restoration is relatively simple in this case but much more involved once learning-rate warm-up is added.

julianmack committed:
Added seq_to_seq state_dict saving

julianmack committed:
Added load_state_dict helper

julianmack committed:
Updated training state to make number of epochs deterministic

julianmack committed:
Updated name of load fn

julianmack committed:
Updated Saver() docstrings

julianmack committed:
Updated documentation

julianmack committed:
Updated lr_update in seq_to_seq.get_state_dict()

julianmack committed:
Updated CallbackHandler Args docstrings

julianmack committed:
Updated docstrings and renamed attributes in CallbackHandler

julianmack committed:
Added fname parsing to extract epoch

samgd commented on src/myrtlespeech/model/seq_to_seq.py:
```diff
@@ -59,3 +59,26 @@ def process(x):
+        """Returns state dict."""
+        state = {
+            "model": self.model.state_dict(),
+            "optim": self.optim.state_dict(),
```
The optimizer is an optional argument (`Optional[torch.optim.Optimizer]`) so can be `None` meaning the call to `self.optim.state_dict()` will throw an `AttributeError`.

julianmack replied:
Ok done!

samgd commented on src/myrtlespeech/model/seq_to_seq.py:
```diff
@@ -59,3 +59,26 @@ def process(x):
+        state = {
+            "model": self.model.state_dict(),
+            "optim": self.optim.state_dict(),
+            "lr_scheduler": self.lr_scheduler.state_dict(),
```
This PR does not add the learning rate scheduler to the SeqToSeq model so the `state_dict` logic should be removed?

julianmack replied:
There is an `lr_scheduler` already in master already (the later PR adds LR warmup and the poly LR that NVIDIA use). It is Optional so I've done as suggested but included checks for `None` for the `lr_scheduler` and `optim`

samgd commented on src/myrtlespeech/model/seq_to_seq.py:
```diff
@@ -59,3 +59,26 @@ def process(x):
 
         return process
+
+    def state_dict(self):
```
PyTorch's state_dict function has the following arguments:

```python
def state_dict(self, destination=None, prefix='', keep_vars=False):
    ....
```

It's not clear that they will ever be used for `SeqToSeq.state_dict` so happy to leave them out pending any use case.

The `state_dict` API also returns a `collections.OrderedDict`, maybe it is worth sticking to this?

Adding the type information should help.

Maybe something like:

```suggestion
    def state_dict(self) -> OrderedDict:
        """Returns a dictionary containing a whole state of the module."""
        state = OrderedDict()
        state["model"] = self.model.state_dict()
        if self.optim is not None:
            state["optim"] = self.optim.state_dict()
        return state
```

Note this removes the `lr_scheduler` and checks for `optim` being None, mentioned below.

julianmack replied:
I've made this into an OrderedDict. 
See above comment on the lr_scheduler being present in master.

julianmack replied:
I've made this into an `OrderedDict`.
See above comment on the `lr_scheduler` being present in master.

samgd commented on src/myrtlespeech/model/seq_to_seq.py:
```diff
@@ -59,3 +59,26 @@ def process(x):
+        }
+        return state
+
+    def load_state_dict(self, state_dict):
```
PyTorch's `load_state_dict` API includes the `strict` parameter:

>             strict (bool, optional): whether to strictly enforce that the keys
>             in :attr:`state_dict` match the keys returned by this module's
>             :meth:`~torch.nn.Module.state_dict` function. Default: ``True``

This is probably worth keeping as, for instance, it may be the case that this needs to be set to `False` as the `optim` doesn't exist for a given checkpoint as it is `Optional` or a transfer learning is being performed so the final FC layer doesn't exist in the model's `state_dict` etc. 

```suggestion
    def load_state_dict(
        self, 
        state_dict: Mapping[str, Mapping[str, torch.Tensor]],
        strict: bool = True
    ) -> Tuple[List[str], List[str]]:
        """See :py:meth:`~torch.nn.Module.load_state_dict`."""
        ...
```

Ensuring to handle the `strict` keyword, check that `optim is not None` and ` remove the `lr_scheduler`?

julianmack replied:
Updated to handle `strict`, and check for `optim` and `lr_scheduler` not None. 

(see above comment re `lr_scheduler`)

samgd reviewed (changes requested):
I suggest making the following changes to Saver/CallbackHandler/fit in order to
make the code more modular by adhering to the separation of concerns design
principle. This will make the code is simpler to modify and maintain in the
future!

Each of the three components has a reasonably well defined purpose and
interface:

    - The Saver (should) handle everything to do with saving and restoring
      state to and from disk.

    - The CallbackHandler is responsible for creating the initial state and
      running the Callbacks at the appropriate times.

    - The fit function orchestrates the training and inference process.

      The fit function uses the CallbackHandler's interface to allow
      modification to this process where the modifications are Callbacks. The
      CallbackHandler uses the Callback interface to run each Callback in turn.

The changes suggested aim to keep these interfaces distinct:

    - The CallbackHandler can initialise the initial_epoch and
      total_train_batches to 0. The `Saver.on_train_begin` method can overwrite
      these values if it has loaded state from disk using the standard
      `Callback` functionality. This means the CallbackHandler's init arguments
      and logic remains unchanged (no need for `epoch` or
      `total_train_batches`).

    - fit does not need to know anything about the `training_state`. The Saver
      Callback is responsible for it and the `for epoch in...` loop becomes
      something like:

      `for epoch in range(cb_handler.state_dict["epoch"], cb_handler.state_dict["epochs"])`

    - Saver needs be able to load state from disk `on_train_begin` and dump
      state `on_epoch_end`.

      In `on_train_begin` the state to load could either be explicitly
      specified in the init function or implicitly chosen based on the last
      state_dict dumped in the `log_dir`. The former is probably simpler for
      now, the latter was useful in the deepspeech repo as we were using
      ephemeral instances that kept restarting.

      `on_epoch_end` essentially becomes `load_seq_to_seq`. It might be worth
      removing the parsing of the epoch from the filename and only allowing it
      to be parsed from the loaded state dict.

julianmack committed:
Removed load_seq_to_seq fn and pushed loading responsibility into Saver callback.

julianmack committed:
Fixed state_dict saving bugs

julianmack committed:
Responded to PR comments

julianmack committed:
Removed strict keyword from optim and lr_scheduler state dict loading

julianmack committed:
Updated docstrings

julianmack committed:
Updated docstrings and removed debug print statement

julianmack commented:
I've followed all suggested changes except for the fact that there is an `lr_scheduler` in master so I've kept this and treated accordingly. 

Putting up for re-review.

samgd commented on src/myrtlespeech/model/seq_to_seq.py:
```diff
@@ -59,3 +61,34 @@ def process(x):
+        self.model.load_state_dict(state_dict["model"], strict=strict)
+
+        if self.optim is not None:
+            self.optim.load_state_dict(state_dict["optim"])
+
+        if self.lr_scheduler is not None:
+            self.lr_scheduler.load_state_dict(state_dict["lr_scheduler"])
```
Should the `strict` argument be passed to these recursive calls to `load_state_dict`?

julianmack replied:
The argument only applies to model state_dict loading:
* `optim`: https://pytorch.org/docs/stable/optim.html#torch.optim.Optimizer.load_state_dict
* `lr_scheduler`: https://pytorch.org/docs/stable/_modules/torch/optim/lr_scheduler.html#_LRScheduler

samgd commented on src/myrtlespeech/run/run.py:
```diff
@@ -170,17 +176,119 @@ def on_train_end(self, **kwargs):
+        for fname in os.listdir(self.log_dir):
+            if re.match(r"state_dict_\d*\.pt", fname):
+                epochs = [int(x) for x in re.findall(r"\d*", fname) if x != ""]
+                fnames.append((epochs[0], fname))
```
`re.match` returns a match object that "always have a boolean value of True. Since match() and search() return None when there is no match, you can test whether there was a match with a simple if statement:"

```python
match = re.search(pattern, string)
if match:
    process(match)
```

https://docs.python.org/3/library/re.html#match-objects

Refactoring to avoid running the regex again in `findall`, perhaps:

```python
for fname in os.listdir(self.log_dir):
    match = re.findall(r"state_dict_(\d+)\.pt", fname)
    if match:
        fnames.append((int(match[1]), fname))
```

This changes the regex slightly to ensure there is an epoch number ("state_dict_.pt" matched previously) and it uses a group to capture the epoch number.

julianmack replied:
Made this change but replacing suggested line `fnames.append((int(match[1]), fname))` -> `fnames.append((int(match[0]), fname))`

samgd commented on src/myrtlespeech/run/run.py:
```diff
@@ -170,17 +176,119 @@ def on_train_end(self, **kwargs):
+            A Tuple of Optional ints ``epoch, total_train_batches``.
+        """
+        fnames = []
+        dict_fname = None
```
This can be deleted? It is unused until being set below when fnames in unpacked?

julianmack replied:
done!

samgd commented on src/myrtlespeech/run/run.py:
```diff
@@ -170,17 +176,119 @@ def on_train_end(self, **kwargs):
+    This callback will also attempt to load the ``state_dict``: from
+    ``load_fp`` if present and otherwise will load the most recent
+    ``state_dict`` in ``log_dir`` (if present) **according to filenames of
+    the form ``state_dict_<EPOCH>.pt``**.
```
This class also loads the model so this should probably be mentioned in the summary line. The docstring also doesn't mention how/where is saves files. Playing with potential changes, maybe something like:

```python
    """Loads and saves the ``model``'s ``state_dict``.

    This callback saves the ``model``'s ``state_dict`` to ``log_dir``
    ``on_epoch_end`` with file name ``state_dict_<EPOCH>.pt``, where
    ``<EPOCH>`` is the number of completed epochs.

    At the start of training, ``on_train_begin``, the callback attempts to load
    the model. If ``load_fp is not None`` it attempts to restore the model from
    the ``state_dict`` saved in ``load_fp``. If ``load_fp is None`` it finds
    the file name in ``log_dir`` with highest ``<EPOCH>`` and attempts to
    restore the model from the ``state_dict`` saved in it.

    Args:
        log_dir: A pathlike object giving the logging directory.

        load_fp: An Optional pathlike object specifying the filepath of the
            state_dict.

        model: A :py:class:`torch.nn.Module` to be saved each epoch. To enable
            resumption of training after a crash, the user should initialise
            this callback with a :py:class:`SeqToSeq` instance.
    """
```

julianmack replied:
Nice. Done.

samgd reviewed (changes requested):
Is the `lr_scheduler` in master? If it is can this PR be rebased on it (I can't see it!). If it isn't, can the `lr_scheduler` code be pushed to a different PR?

Now that the `Saver` functionality is more complex should we add a test(s)? A simple create tiny module, create temporary directory, run a bunch of epochs with a tiny amount of data, make sure the right state_dict is restored?

julianmack commented:
> Is the lr_scheduler in master? If it is can this PR be rebased on it (I can't see it!). If it isn't, can the lr_scheduler code be pushed to a different PR?

I think my first comment on the `lr_scheduler` was unclear @samgd. There is already a `lr_scheduler` in master: https://github.com/MyrtleSoftware/myrtlespeech/blob/d4a0003197f4512eb4ca9b1cda74358456a77990/src/myrtlespeech/builders/task_config.py#L94

which was added by Giuseppe a while back. This does not include my changes which add `lr_warmup` and `polynomial_lr`. These latter changes are in an incoming PR (https://github.com/MyrtleSoftware/myrtlespeech/pull/22) and I agree that these should be considered separately.

I will respond to the rest of the above comments shortly but just flagging this high level point now

samgd commented:
> I think my first comment on the lr_scheduler was unclear @samgd. There is already a lr_scheduler in master but not added by me (added by Giuseppe a while back):

Ah I see now, thanks. The `SeqToSeq` class not having an `lr_scheduler` parameter in __init__ or documented attribute was what was throwing me off - I had forgotten it was being set in the builder. I'll make an issue for this as it should be explicit and obvious!

julianmack committed:
Added Saver test

julianmack committed:
refactored Saver test to make individual tests quicker

julianmack committed:
Small changes according to samgd PR comments

julianmack committed:
Updated Saver docstrings

julianmack commented:
Responded to changes and added a test.

I've added a `check_state_dicts_match` function in `test.utils.util`. Although this is currently used in just the new test, I've placed it here because I want to use it in another test in my incoming lr_warmup PR. 
I've also disabled the  `builders/test_task_config.py`` test as this functionality is necessarily tested in at the start of the new `Saver` test.

julianmack commented:
> Ah I see now, thanks. The `SeqToSeq` class not having an `lr_scheduler` parameter in **init** or parameter documented in the docstring was what was throwing me off - I had forgotten it was being set in the builder. I'll make an issue for this as it should be explicit and obvious!

Also on the above point - in the `lr_warmup` PR, I've refactored the `lr_sheduler` creation and added a proper builder w. sphinx docs.

samgd commented on src/myrtlespeech/run/run.py:
```diff
@@ -170,17 +176,123 @@ def on_train_end(self, **kwargs):
+        """
+        fnames = []
+        for fname in os.listdir(self.log_dir):
+            match = re.findall(r"state_dict_(\d+)\.pt", fname)
```
The following is my mistake, sorry, but `findall` should actually be `fullmatch`:

```suggestion
            match = re.fullmatch(r"state_dict_(\d+)\.pt", fname)
```

julianmack replied:
I think I actually want the `findall` version since this specifically returns the epoch number (i.e. it matches the whole regex but returns the matches in parentheses). 

with this I can do:
```python
if match:
    fnames.append((int(match[0]), fname))
fnames.sort()
```
And I later check that the epoch in the filename is the same as the epoch in the state dict and raise a warning if not.

samgd commented on tests/run/test_Saver.py:
```diff
@@ -0,0 +1,113 @@
+# Utilities -------------------------------------------------------------------
+
+
+def change_seq_to_seq_state_dict(seq_to_seq: SeqToSeq):
```
Name is a little opaque, could make clearer if you want to update or have any suggestions (I can't think of any, naming is hard!). Name change aside, perhaps adding an underscore to signify in-place and updating the top level description? 

```suggestion
def change_seq_to_seq_state_dict_(seq_to_seq: SeqToSeq) -> None:
    """Modifies all values in ``seq_to_seq`` such that they are different"""
```

julianmack replied:
Changed to `alter_seq2seq_attributes_`?

samgd commented on tests/builders/test_task_config.py:
```diff
@@ -1,22 +1,24 @@
-def test_build_returns(task_cfg: task_config_pb2.TaskConfig) -> None:
-    """Test that build returns when called."""
-    model, epochs, train_loader, eval_loader = build(task_cfg)
-    assert isinstance(model, torch.nn.Module)
```
It's probably worth keeping this test such that there is a minimal test case for building the task config? It's duplicated in the Saver test but this could be modified in future or fail in any of the later parts after the build.

julianmack replied:
Ok I've added back in (but keeping the helper function)

samgd commented on tests/run/test_Saver.py:
```diff
@@ -0,0 +1,113 @@
+        seq_to_seq_cfg
+    )
+
+    with TemporaryDirectory() as tmpdir:
```
pytest has a built-in fixture for temporary directories: https://docs.pytest.org/en/latest/tmpdir.html

julianmack replied:
I was using `tmpdir` originally but it interacts with hypothesis in an undesirable way. Namely,  a single `tmpdir` is shared between all hypothesis calls to the same function. In this case that causes the test to fail when it shouldn't as the directory contents affect `Saver`s loading behaviour.  

I could add teardown code but I think that removes the simplicity that `tmpdir` usually provides.

samgd commented on tests/run/test_Saver.py:
```diff
@@ -0,0 +1,113 @@
+                check_state_dicts_match(
+                    expected_sd["lr_scheduler"],
+                    seq_to_seq.state_dict()["lr_scheduler"],
+                )
```
The `check_state_dicts_match` function is recursive so these three calls could be merged into one? The `check_state_dicts_match` function could also return a bool rather than throwing an assertion making it more natural to call?

```python
assert state_dicts_match(expected_sd, seq_to_seq.state_dict())
```
    
with `check_state_dicts_match` becoming something like (untested):

```python
def state_dicts_match(dict1, dict2) -> bool:
    """Returns True if dicts have same keys and values."""
    if not dict1.keys() == dict2.keys():
        return False
    for key in dict1.keys():
        val1 = dict1[key]
        val2 = dict2[key]
        if isinstance(val1, dict):
            if not (isinstance(val2, dict) and state_dicts_match(val1, val2)):
                return False
        if isinstance(val1, float):
            if not (isinstance(val2, float) and math.isclose(val1, val2)):
                return False
        if isinstance(val1, torch.Tensor):
            if not (isinstance(val2, torch.Tensor) and torch.allclose(val1, val2)):
               return False
    return True
```

julianmack replied:
I've updated it to return a bool but I'm intentionally checking that *all* of the subdicts don't match whereas `assert state_dicts_match(expected_sd, seq_to_seq.state_dict())` will fail if just one of them was changed.

samgd commented on tests/run/test_Saver.py:
```diff
@@ -0,0 +1,113 @@
+                callbacks=[Saver(log_dir=tmpdir, model=seq_to_seq)],
+            )
+
+            check_state_dicts_match(expected_sd, seq_to_seq.state_dict())
```
Are both of these required? `fit` does the `not check_with_fit_api` branch internally?

julianmack replied:
Sure thing- I thought it was useful to check that both work but it is probably overkill! I've removed the first branch.

samgd commented on tests/run/test_Saver.py:
```diff
@@ -0,0 +1,113 @@
+# Tests -----------------------------------------------------------------------
+
+
+@settings(deadline=3000)
```
Is this deadline required? Can we reduce this by a lot? 3s/test is long, the models only really need to contain 1 parameter

julianmack replied:
removed

samgd commented on tests/utils/utils.py:
```diff
@@ -93,3 +93,25 @@ def tensors(
+TOL = 1e-8
+
+
+def check_state_dicts_match(dict1, dict2):
```
See `test_Saver.py` comment about refactoring this.

samgd commented on tests/run/test_Saver.py:
```diff
@@ -0,0 +1,113 @@
+import copy
+from tempfile import TemporaryDirectory
+
```
`test_Saver.py` -> `test_saver.py`?

samgd reviewed (changes requested):
Minor comments, but nearly there!

julianmack committed:
Altered test helper function naming

julianmack committed:
Re-added task_config test

julianmack committed:
Updated state_dict_match function

julianmack committed:
Removed one of the Saver test branches

julianmack committed:
Renamed test for Saver

julianmack commented:
Responded to comments. A few outstanding discussions.

samgd committed:
Fix ModelCallback bug and train on full LibriSpeech
